File: src/two_quadrotors_lf.cpp
/*
@description:
This node executes the controller in order to solve the formation control problem, using the leader-follower 
scheme, with two simulated drones. The leader drone tracks its desired trajectory (using the velocity 
controller), and then the follower drone follows the offset leader's trajectory.
For this example, a helix trajectory is used.

-- To land the drone and finish this node, please press 'Esc' key. Do not use ctrl - C to finish this node

DIRECTIONS:
In different terminals:
$ ros2 launch my_uavs two_ign_world_launch.py
	# Launchs two simulated quadrotors in Ignition, runs ros_gz_bridge and opens Rviz

$ ros2 run my_uavs two_quadrotors_lf r1 r2
	# Runs this node. NOTE: Do not run this node within a launch file
*/

#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "nav_msgs/msg/odometry.hpp"
// #include "nav_msgs/msg/path.hpp"
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>

//Includes needed by keyboard events
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

using namespace std::chrono_literals;
using std::placeholders::_1;

//Create a custom structure to define the robot posture (position and orientation, using Euler angles)
struct robot_pose{ double x, y, z, roll, pitch, yaw; };

double takeoff_alt = 1.2; //Desired takeoff altitude
int key;

//Trajectory variables
double Vxy_max = 1.5, Vz_max = 0.5, Wz_max = 4; //maximum translational and rotational velocities, in [m/s] and [rad/s]
double T = 50, k; //Trajectory period, controller gains kx = ky = k

//Follower drone variables
double s_d = 1.5, alp_d = M_PI, s_zd = 0;


int kbhit(void){ //kbhit function
  struct termios oldt, newt;
  int ch, oldf;

  tcgetattr(STDIN_FILENO, &oldt);
  newt = oldt;
  newt.c_lflag &= ~(ICANON | ECHO);
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
  fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);

  ch = getchar();

  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
  fcntl(STDIN_FILENO, F_SETFL, oldf);

  if(ch != EOF)
  {
    ungetc(ch, stdin);
    return 1;
  }
  return 0;
}

class LFController : public rclcpp::Node
{
  public:
    LFController( char * argv[] ) : Node("two_quadrotors_lf")
    {
      ns1_ = argv[1];
      ns2_ = argv[2];

      publisher1_ = this->create_publisher<geometry_msgs::msg::Twist>(ns1_+"/cmd_vel", 2);
      publisher2_ = this->create_publisher<geometry_msgs::msg::Twist>(ns2_+"/cmd_vel", 2);

      timer_ = this->create_wall_timer(20ms, std::bind(&LFController::loop, this)); //20ms = 50 Hz
      
      subscriber1_ = this->create_subscription<nav_msgs::msg::Odometry>(ns1_+"/odom", 
            2, std::bind(&LFController::odom_callback1, this, _1));
      
      subscriber2_ = this->create_subscription<nav_msgs::msg::Odometry>(ns2_+"/odom", 
            2, std::bind(&LFController::odom_callback2, this, _1));
      
      RCLCPP_INFO(this->get_logger(), "Press 'ESC' key to land the drones and to finish the node");
    }
    
    ~LFController() //Class destructor definition
    {
      RCLCPP_INFO(this->get_logger(), "Node has been terminated"); 
    }
    
    void odom_callback1(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
      r_pose1_.x = msg->pose.pose.position.x;
      r_pose1_.y = msg->pose.pose.position.y;
      r_pose1_.z = msg->pose.pose.position.z;
      
      //GET THE EULER ANGLES FROM THE QUATERNION
      tf2::Quaternion q( msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
                    msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);

      // 3x3 Rotation matrix from quaternion
      tf2::Matrix3x3 m(q);
      // Roll, Pitch and Yaw from rotation matrix
      double roll, pitch;
      m.getRPY(roll, pitch, r_pose1_.yaw); //Euler angles in rad
    }

    void odom_callback2(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
      r_pose2_.x = msg->pose.pose.position.x;
      r_pose2_.y = msg->pose.pose.position.y;
      r_pose2_.z = msg->pose.pose.position.z;
      
      //GET THE EULER ANGLES FROM THE QUATERNION
      tf2::Quaternion q( msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
                    msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);

      // 3x3 Rotation matrix from quaternion
      tf2::Matrix3x3 m(q);
      // Roll, Pitch and Yaw from rotation matrix
      double roll, pitch;
      m.getRPY(roll, pitch, r_pose2_.yaw); //Euler angles in rad
    }

    void movement1(float x, float y, float z, float turn)
    { //Function to assign control signals (Vx, Vy, Vz, Wz)
      vel_msg1_.linear.x = x;   vel_msg1_.linear.y = y;   vel_msg1_.linear.z = z;   vel_msg1_.angular.z = turn;
    }

    void movement2(float x, float y, float z, float turn)
    { //Function to assign control signals (Vx, Vy, Vz, Wz)
      vel_msg2_.linear.x = x;   vel_msg2_.linear.y = y;   vel_msg2_.linear.z = z;   vel_msg2_.angular.z = turn;
    }

    void leader_vel_control() //Function to generate the desired trajectory and to compute the signals control of the leader)
    { 
      double Vx, Vy, Vz, Wz; //Define control signals of the drone
      double ex, ey, ez, e_yaw; //Tracking errors
      
      //Desired positions, desired orientation and time derivative, respectively
      double Xd, Yd, Zd, Yawd, Xdp, Ydp, Zdp, Yawdp; 

      //Desired trajectory: Helix
      double X0 = 0, Y0 = 0, Z0 = takeoff_alt, radius = 3, w = 2*M_PI/T, Vzd = 0.02;
      
      double t = now().seconds()-t0_; //Compute the time since this node started

      //Desired position in the 3D space
      Xd = X0+radius*sin(w*t);
      Yd = Y0+radius*cos(w*t);
      Zd = Z0+Vzd*t; //linear increasing
      Yawd = sin(w*t); //sinusoidal behavior

      //Corresponding time derivatives
      Xdp = radius*w*cos(w*t);
      Ydp = -radius*w*sin(w*t);
      Zdp = Vzd;
      Yawdp = w*cos(w*t);

      ex = r_pose1_.x-Xd; ey = r_pose1_.y-Yd; ez = r_pose1_.z-Zd; e_yaw = r_pose1_.yaw - Yawd; //Compute tracking errors

      //Kinematic controller. Auxiliary controls, in global coordinates
      double Ux = Xdp-k*ex;
      double Uy = Ydp-k*ey;
  
      //Translational velocities with respect to the robot frame (rotation around Z axis)
      Vx = Ux*cos(r_pose1_.yaw)+Uy*sin(r_pose1_.yaw);
      Vy = -Ux*sin(r_pose1_.yaw)+Uy*cos(r_pose1_.yaw);

      //Kinematic controller. Note: Vz and Wz are compute directly since they are given with respect to the robot frame
      Vz = Zdp-k*ez;
      Wz = Yawdp-k*e_yaw;

      //Velocities saturation
      if( abs(Vx)>Vxy_max ){ Vx = Vxy_max*abs(Vx)/Vx; }
      if( abs(Vy)>Vxy_max ){ Vy = Vxy_max*abs(Vy)/Vy; }
      if( abs(Vz)>Vz_max ){ Vz = Vz_max*abs(Vz)/Vz; }
      if( abs(Wz)>Wz_max ){ Wz = Wz_max*abs(Wz)/Wz; }

      movement1(Vx,Vy,Vz,Wz);
      //Assignation of leader velocities required by the follower robot
      Vxl_ = Vx;
      Vyl_ = Vy;
      Vzl_ = Vz;
      Wzl_ = Wz;
    }

    void follower_vel_control() //Function to compute the signals control of the follower robot
    { 
      //Note: desired follower position and yaw orientation are given by the leader's ones
      double Vxf, Vyf, Vzf, Wzf;
      double exf, eyf, ezf, e_yawf;
      double Xfd, Yfd, Xfdp, Yfdp, xlp, ylp, Ux, Uy;

      //Follower desired position with respect to the global frame
      Xfd = r_pose1_.x+s_d*cos(r_pose1_.yaw+alp_d);
      Yfd = r_pose1_.y+s_d*sin(r_pose1_.yaw+alp_d);

      //Translational velocities of the leader drone in the global frame
      xlp = Vxl_*cos(r_pose1_.yaw)-Vyl_*sin(r_pose1_.yaw);
      ylp = Vxl_*sin(r_pose1_.yaw)+Vyl_*cos(r_pose1_.yaw);

      //Corresponding desired time derivatives
      Xfdp = xlp-Wzl_*s_d*sin(r_pose1_.yaw+alp_d);
      Yfdp = ylp+Wzl_*s_d*cos(r_pose1_.yaw+alp_d);

      //Compute tracking errors (with respect to the global frame)
      exf = r_pose2_.x-Xfd;
      eyf = r_pose2_.y-Yfd; 
      ezf = r_pose2_.z-(r_pose1_.z+s_zd);  //Both the altitude and the orientation are equal to leader drone's ones
      e_yawf = r_pose2_.yaw - r_pose1_.yaw;

      //Kinematic controller. Auxiliary controls, in global coordinates
      Ux = Xfdp-k*exf; 	Uy = Yfdp-k*eyf;
      
      //Translational velocities with respect to the robot frame
      Vxf = Ux*cos(r_pose2_.yaw)+Uy*sin(r_pose2_.yaw);
      Vyf = -Ux*sin(r_pose2_.yaw)+Uy*cos(r_pose2_.yaw);

      //Vertical and rotational velocities
      Vzf = Vzl_-k*ezf;
      Wzf = Wzl_-k*e_yawf;

      //Velocities saturation
      if( abs(Vxf)>Vxy_max ){Vxf = Vxy_max*abs(Vxf)/Vxf; }
      if( abs(Vyf)>Vxy_max ){Vyf = Vxy_max*abs(Vyf)/Vyf; }
      if( abs(Vzf)>Vz_max ){Vzf = Vz_max*abs(Vzf)/Vzf; }
      if( abs(Wzf)>Wz_max ){Wzf = Wz_max*abs(Wzf)/Wzf; }

      movement2(Vxf,Vyf,Vzf,Wzf);
}

    void loop()
    {
      if(mode_ == 0)
      {
        RCLCPP_WARN_ONCE(this->get_logger(), "Desired takeoff altitude = %.2f", takeoff_alt);
        RCLCPP_INFO_ONCE(this->get_logger(), "Taking off ...");

        if(r_pose1_.z >= takeoff_alt || r_pose2_.z >= takeoff_alt){
          mode_ = 1;
          t0_ = now().seconds(); //Get the initial time for tracking
        }else{
          movement1(0,0,0.2,0);
          movement2(0,0,0.2,0);
        } 
      }

      if(mode_ == 1) //leader-follower scheme
      {
        RCLCPP_INFO_ONCE(this->get_logger(), "Tracking trajectory ...");
        leader_vel_control();
        follower_vel_control();
      }

      if(mode_ == 2){ //land
        RCLCPP_INFO_ONCE(this->get_logger(), "Landing ...");
        movement1(0,0,-0.2,0);
        movement2(0,0,-0.2,0);
        if(r_pose1_.z < 0.2) movement1(0,0,0,0); //Stop robot to wait for the other
        if(r_pose2_.z < 0.2) movement2(0,0,0,0);
        
        if(r_pose1_.z < 0.2 && r_pose2_.z < 0.2) rclcpp::shutdown(); //Terminate this node
      }
      
      //Display some variables of interest
      if(freq_div_ >= 25 && mode_ == 1)
      {
        // RCLCPP_INFO(this->get_logger(), "");
        freq_div_ = 0;
        
      }else freq_div_++;

      //Check for any key press
      if(kbhit())	key = getchar();
      if(key == 27) mode_ = 2; //ESC key = 27 in Ascii code

      publisher1_->publish(vel_msg1_); //Publish robot velocities
      publisher2_->publish(vel_msg2_);
    }

  private:
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher1_, publisher2_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr subscriber1_, subscriber2_;
   
    geometry_msgs::msg::Twist vel_msg1_, vel_msg2_; //Create a msg obj to publish velocity
    double Vxl_, Vyl_, Vzl_, Wzl_; //Signals control of the leader required by the follower

    std::string ns1_, ns2_; 
    
    int freq_div_, mode_ = 0; //mode = 0 taking off, 1 trajectory tracking, 2 landing
    robot_pose r_pose1_, r_pose2_; //drones posture variables
    double t0_;
};

int main(int argc, char * argv[])
{
  auto logger = rclcpp::get_logger("logger");

  RCLCPP_INFO(logger, "argc = %d", argc);
  // Obtain parameters from command line arguments

  if (argc < 3) //node_name and ns1, ns2 = 3; if this node is run within a launcher, node_name and ns1, ns2 --ros-args = 4
  {
    RCLCPP_ERROR(logger, "Robots namespace not given\nUsage: "
      "$ ros2 run my_uavs two_quadrotors_lf r1 r2");
    return 1;
  }

  rclcpp::init(argc, argv);
  auto node = std::make_shared<LFController>(argv);
  rclcpp::spin(node);

  rclcpp::shutdown();
  return 0;
}

File: scripts/two_quadrotors_lf.py
#!/usr/bin/env python3

"""
@description:
This node executes the controller in order to solve the formation control problem, using the leader-follower 
scheme, with two simulated drones. The leader drone tracks its desired trajectory (using the velocity 
controller), and then the follower drone follows the offset leader's trajectory.
For this example, a helix trajectory is used.
This node also publish transformation between robots frame.

-- To land the drone and finish this node, please press 'Esc' key. Do not use ctrl - C to finish this node

DIRECTIONS:
In different terminals:
$ ros2 launch my_uavs two_ign_world_launch.py
	# Launchs two simulated quadrotors in Ignition, runs ros_gz_bridge and opens Rviz

$ ros2 run my_uavs two_quadrotors_lf.py r1 r2
	# Runs this node. NOTE: Do not run this node within a launch file
"""

__author__ = "David Caceres"
__contact__ = "davidxcaceres1@gmail.com"

import rclpy, os, select, sys, math
from math import pi
from rclpy.node import Node
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry

from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster

#To work with Euler angles and quaternions conversions
import tf_transformations

"""
It is required to install:
$ sudo apt install ros-$ROS_DISTRO-tf-transformations
$ sudo apt install python3-pip
$ sudo pip3 install transforms3d #Python package
"""

takeoff_alt = 1.2; #Desired takeoff altitude

if os.name == 'nt':
    import msvcrt
else:
    import termios
    import tty

key = 0

Vxy_max = 1.5; Vz_max = 0.5; Wz_max = 4 #maximum translational and rotational velocities, in [m/s] and [rad/s]
T = 50.0 #Trajectory period
k = 0.4 #controller gains kx = ky = k

#Follower drone variables
s_d = 1.5; alp_d = pi; s_zd = 0.0


def get_key(settings):
    if os.name == 'nt':
        return msvcrt.getch().decode('utf-8')
    tty.setraw(sys.stdin.fileno())
    rlist, _, _ = select.select([sys.stdin], [], [], 0.1)
    if rlist:
        key = sys.stdin.read(1)
    else:
        key = ''

    termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
    return key


class LFController(Node):

    def __init__(self, argv):
        super().__init__('two_quadrotors_lf') # class’s constructor

        self.ns1 = argv[1]
        self.ns2 = argv[2]

        self.publisher1 = self.create_publisher(Twist, self.ns1+'/cmd_vel', 2)
        self.publisher2 = self.create_publisher(Twist, self.ns2+'/cmd_vel', 2)

        self.timer = self.create_timer(0.02, self.loop) # 20ms = 50 Hz

        self.subscription1 = self.create_subscription(Odometry, self.ns1+'/odom',self.odom_callback1, 2)
        self.subscription2 = self.create_subscription(Odometry, self.ns2+'/odom',self.odom_callback2, 2)

        self.vel_msg1 = Twist()
        self.vel_msg2 = Twist()

        self.freq_div = 0
        self.mode = 0 #mode = 0 taking off, 1 moving around, 2 landing

        #Define the robots' posture using lists [x,y,z,roll,pitch,yaw]
        self.r_pose1 = [0.0,0.0,0.0,0.0,0.0,0.0] 
        self.r_pose2 = [0.0,0.0,0.0,0.0,0.0,0.0]

        #Signals control of the leader required by the follower
        self.Vxl = 0.0;  self.Vyl = 0.0; self.Vzl = 0.0; self.Wzl = 0.0 

        self.settings = None
        if os.name != 'nt':
            self.settings = termios.tcgetattr(sys.stdin)


        self.frame_id = '/odom'
        self.child_frame_id = '/base_footprint'
        
        # Initialize the transform broadcaster
        self.tf_broadcaster1 = TransformBroadcaster(self)
        self.tf_broadcaster2 = TransformBroadcaster(self)
    
        self.get_logger().info('Press "ESC" to land the drone and to finish the node')


    def __del__(self):
        self.get_logger().info('Node has been terminated')
        

    def odom_callback1(self, msg):
        self.r_pose1[0] = msg.pose.pose.position.x
        self.r_pose1[1] = msg.pose.pose.position.y
        self.r_pose1[2] = msg.pose.pose.position.z

        #GET THE EULER ANGLES FROM QUATERNION
        q = [0.0,0.0,0.0,1.0] #Define a unitary quaternion. Important: use [] not ()

        q[0] = msg.pose.pose.orientation.x
        q[1] = msg.pose.pose.orientation.y
        q[2] = msg.pose.pose.orientation.z
        q[3] = msg.pose.pose.orientation.w

        self.r_pose1[3], self.r_pose1[4], self.r_pose1[5] = tf_transformations.euler_from_quaternion(q)


        # Create and fill out the transform broadcaster
        tranStamp = TransformStamped()

        tranStamp.header.stamp = self.get_clock().now().to_msg()
        tranStamp.header.frame_id = self.ns1+self.frame_id
        tranStamp.child_frame_id = self.ns1+self.child_frame_id

        tranStamp.transform.translation.x = self.r_pose1[0]
        tranStamp.transform.translation.y = self.r_pose1[1]
        tranStamp.transform.translation.z = self.r_pose1[2]

        tranStamp.transform.rotation.x = q[0]
        tranStamp.transform.rotation.y = q[1]
        tranStamp.transform.rotation.z = q[2]
        tranStamp.transform.rotation.w = q[3]

        self.tf_broadcaster1.sendTransform(tranStamp) # Send the transformation
    

    def odom_callback2(self, msg):
        self.r_pose2[0] = msg.pose.pose.position.x
        self.r_pose2[1] = msg.pose.pose.position.y
        self.r_pose2[2] = msg.pose.pose.position.z

        #GET THE EULER ANGLES FROM QUATERNION
        q = [0.0,0.0,0.0,1.0] #Define a unitary quaternion. Important: use [] not ()

        q[0] = msg.pose.pose.orientation.x
        q[1] = msg.pose.pose.orientation.y
        q[2] = msg.pose.pose.orientation.z
        q[3] = msg.pose.pose.orientation.w

        self.r_pose2[3], self.r_pose2[4], self.r_pose2[5] = tf_transformations.euler_from_quaternion(q)


        # Create and fill out the transform broadcaster
        tranStamp = TransformStamped()

        tranStamp.header.stamp = self.get_clock().now().to_msg()
        tranStamp.header.frame_id = self.ns2+self.frame_id
        tranStamp.child_frame_id = self.ns2+self.child_frame_id

        tranStamp.transform.translation.x = self.r_pose2[0]
        tranStamp.transform.translation.y = self.r_pose2[1]
        tranStamp.transform.translation.z = self.r_pose2[2]

        tranStamp.transform.rotation.x = q[0]
        tranStamp.transform.rotation.y = q[1]
        tranStamp.transform.rotation.z = q[2]
        tranStamp.transform.rotation.w = q[3]

        self.tf_broadcaster2.sendTransform(tranStamp) # Send the transformation


    def movement1(self,x,y,z,turn):
        self.vel_msg1.linear.x = x
        self.vel_msg1.linear.y = y
        self.vel_msg1.linear.z = z
        self.vel_msg1.angular.z = turn

    def movement2(self,x,y,z,turn):
        self.vel_msg2.linear.x = x
        self.vel_msg2.linear.y = y
        self.vel_msg2.linear.z = z
        self.vel_msg2.angular.z = turn

    
    def leader_vel_control(self): #Function to generate the desired trajectory and to compute the signals control
        global T, k, takeoff_alt #Indicate that some variables are global
        #Desired trajectory: Helix
        X0 = 0.0; Y0 = 0.0; Z0 = takeoff_alt; radius = 3.0; w = 2*pi/T; Vzd = 0.02

        #Compute the time since this node started
        t = self.get_clock().now().nanoseconds / 1e9 - self.t0

        #Desired position in the 3D space
        Xd = X0+radius*math.sin(w*t)
        Yd = Y0+radius*math.cos(w*t)
        Zd = Z0+Vzd*t #linear increasing
        Yawd = math.sin(w*t) #sinusoidal behavior

        #Corresponding time derivatives
        Xdp = radius*w*math.cos(w*t)
        Ydp = -radius*w*math.sin(w*t)
        Zdp = Vzd
        Yawdp = w*math.cos(w*t)

        #Compute tracking errors
        ex = self.r_pose1[0]-Xd; ey = self.r_pose1[1]-Yd; ez = self.r_pose1[2]-Zd
        e_yaw = self.r_pose1[5] - Yawd; 

        #Kinematic controller. Auxiliary controls, in global coordinates
        Ux = Xdp-k*ex
        Uy = Ydp-k*ey
  
        #Translational velocities with respect to the robot frame (rotation around the Z axis)
        Vx = Ux*math.cos(self.r_pose1[5])+Uy*math.sin(self.r_pose1[5])
        Vy = -Ux*math.sin(self.r_pose1[5])+Uy*math.cos(self.r_pose1[5])

        #Kinematic controller. Note: Vz and Wz are compute directly since they are given with respect to the robot frame
        Vz = Zdp-k*ez
        Wz = Yawdp-k*e_yaw

        #Velocities saturation
        if abs(Vx)>Vxy_max:
            Vx = Vxy_max*abs(Vx)/Vx
        if abs(Vy)>Vxy_max:
            Vy = Vxy_max*abs(Vy)/Vy
        if abs(Vz)>Vz_max:
            Vz = Vz_max*abs(Vz)/Vz
        if abs(Wz)>Wz_max:
            Wz = Wz_max*abs(Wz)/Wz

        self.movement1(Vx,Vy,Vz,Wz)

        self.Vxl = Vx
        self.Vyl = Vy
        self.Vzl = Vz
        self.Wzl = Wz


    def follower_vel_control(self): #Function to compute the signals control of the follower robot
        #Note: desired follower position and yaw orientation are given by the leader's ones
        global k, s_d, alp_d, s_zd        

        #Follower desired position with respect to the global frame
        Xfd = self.r_pose1[0]+s_d*math.cos(self.r_pose1[5]+alp_d)
        Yfd = self.r_pose1[1]+s_d*math.sin(self.r_pose1[5]+alp_d)

        #Translational velocities of the leader drone in the global frame
        xlp = self.Vxl*math.cos(self.r_pose1[5])-self.Vyl*math.sin(self.r_pose1[5])
        ylp = self.Vxl*math.sin(self.r_pose1[5])+self.Vyl*math.cos(self.r_pose1[5])

        #Corresponding desired time derivatives
        Xfdp = xlp-self.Wzl*s_d*math.sin(self.r_pose1[5]+alp_d)
        Yfdp = ylp+self.Wzl*s_d*math.cos(self.r_pose1[5]+alp_d)

        #Compute tracking errors (with respect to the global frame)
        exf = self.r_pose2[0]-Xfd
        eyf = self.r_pose2[1]-Yfd; 
        ezf = self.r_pose2[2]-(self.r_pose1[2]+s_zd) #Both the altitude and the orientation are equal to leader drone's ones
        e_yawf = self.r_pose2[5] - self.r_pose1[5]

        #Kinematic controller. Auxiliary controls, in global coordinates
        Ux = Xfdp-k*exf; 	Uy = Yfdp-k*eyf
      
        #Translational velocities with respect to the robot frame
        Vxf = Ux*math.cos(self.r_pose2[5])+Uy*math.sin(self.r_pose2[5])
        Vyf = -Ux*math.sin(self.r_pose2[5])+Uy*math.cos(self.r_pose2[5])

        #Vertical and rotational velocities
        Vzf = self.Vzl-k*ezf
        Wzf = self.Wzl-k*e_yawf
    
        #Velocities saturation
        if abs(Vxf)>Vxy_max:
            Vxf = Vxy_max*abs(Vxf)/Vxf
        if abs(Vyf)>Vxy_max:
            Vyf = Vxy_max*abs(Vyf)/Vyf
        if abs(Vzf)>Vz_max:
            Vzf = Vz_max*abs(Vzf)/Vzf
        if abs(Wzf)>Wz_max:
            Wzf = Wz_max*abs(Wzf)/Wzf

        self.movement2(Vxf,Vyf,Vzf,Wzf)


    def loop(self):
        if self.mode == 0:
            self.get_logger().info('Desired takeoff altitude = %.2f\n'
                            ' Taking off ...' % takeoff_alt, once=True)
            if self.r_pose1[2] >= takeoff_alt or self.r_pose2[2] >= takeoff_alt:
                self.mode = 1
                self.t0 = self.get_clock().now().nanoseconds / 1e9 #Get the initial time for tracking
            else:
                self.movement1(0.0,0.0,0.2,0.0)
                self.movement2(0.0,0.0,0.2,0.0)
        
        if self.mode == 1: #trajectory tracking
            self.get_logger().info('Trajectory tracking ...', once=True)
            self.leader_vel_control()
            self.follower_vel_control()
        
        if self.mode == 2: #Land
            self.get_logger().info('Landing ...', once=True)
            self.movement1(0.0,0.0,-0.2,0.0)
            self.movement2(0.0,0.0,-0.2,0.0)

            if self.r_pose1[2] < 0.2: #Stop robot to wait for the other
                self.movement1(0.0,0.0,0.0,0.0)

            if self.r_pose2[2] < 0.2:
                self.movement2(0.0,0.0,0.0,0.0)
            
            if self.r_pose1[2] < 0.2 and self.r_pose2[2] < 0.2:
                self.destroy_node()
                sys.exit(0) #Required to release the terminal

        #Display some variables of interest
        if self.freq_div >= 25 and self.mode == 1:
            # self.get_logger().info('' %())
            self.freq_div = 0
        else:
            self.freq_div = self.freq_div + 1
        
        #Check any key press
        key = get_key(self.settings)
        if key ==  '\x1b' or key == 'q' or key == 'Q': #ESC key = '\x1b'
            self.mode = 2
        if key == '\x03': #Ctrl - C
            print("Ctrl-C detected, signal_handler(signum=2)")
            self.destroy_node()
            sys.exit(0)

        self.publisher1.publish(self.vel_msg1) #Publish the 4 control signals
        self.publisher2.publish(self.vel_msg2)
  

def main(args=None): 
    logger = rclpy.logging.get_logger('logger')

    logger.info('argc = %d' % len(sys.argv))

    #Obtain parameters from command line arguments
    if len(sys.argv) < 3: #node_name and ns1, ns2 = 3; if this node is run within a launcher, node_name and ns1, ns2 --ros-args = 4
        logger.error('Robots namespace not given\nUsage: $ ros2 run my_uavs two_quadrotors_lf.py r1 r2')
        sys.exit(1)
  
    rclpy.init(args=args)
    node = LFController(sys.argv)
    
    rclpy.spin(node) #Terminal blocked here


if __name__ == '__main__':
    main()

